src/Microsoft.AspNet.Server.Kestrel/Http/Connection.cs

```diff
@@ -63,13 +65,53 @@ public Connection(ListenerContext context, UvStreamHandle socket) : base(context
             ConnectionControl = this;
         }
 
+        private void ExtractIP()
+        {
+            // storage for binary IP addresses
+            Libuv.sockaddr peer;
+            Libuv.sockaddr local;
+            // length of sockaddr structs
+            var peerLength = Marshal.SizeOf(typeof(Libuv.sockaddr));
+            var localLength = peerLength;
+            // buffers for addresses
+            var peerString = new StringBuilder(50);
+            var localString = new StringBuilder(50);
+            // storage for errors during decoding processes
+            Exception peerException;
+            Exception localException;
+
+            // get local and peer IPs in binary form
+            _socket.Libuv.tcp_getsockname((UvTcpHandle)_socket, out local, ref localLength);
+            _socket.Libuv.tcp_getpeername((UvTcpHandle)_socket, out peer, ref peerLength);
+            // decode as IPv4 addresses
+            _socket.Libuv.ip4_name(ref local, localString, 50, out localException);
```

Comment:
Yes, that's the whole point of dual-stack sockets. In libuv, the OS's defaults are harmonized and you have to explicitly disable (!) ([see uv_tcp_bind](http://docs.libuv.org/en/latest/tcp.html#c.uv_tcp_bind)) dual-stack support

Here is an excerpt from my Azure virtual machine running Ubuntu. You can see that port 5004 is only opened as `tcp6`, but the two HEAD requests using IPv4 and IPv6 respectively both succeed:
```
$ netstat -polta
(Not all processes could be identified, non-owned process info
 will not be shown, you would have to be root to see it all.)
Active Internet connections (servers and established)
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name Timer
tcp        0      0 *:ssh                   *:*                     LISTEN      -                off (0.00/0/0)
[redacted]
tcp        0    556 [redacted]:ssh          [redacted]              ESTABLISHED -                on (0.06/0/0)
tcp6       0      0 [::]:5004               [::]:*                  LISTEN      26133/mono       off (0.00/0/0)
tcp6       0      0 [::]:ssh                [::]:*                  LISTEN      -                off (0.00/0/0)
$ curl --head http://127.0.0.1:5004/
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Connection: close

$ curl --head http://[::1]:5004/
HTTP/1.1 200 OK
Content-Type: text/html; charset=utf-8
Connection: close

$
```